crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -216,74 +216,139 @@ impl<'db> InnerIntersectionBuilder<'db> {
     }
 
     /// Adds a positive type to this intersection.
-    fn add_positive(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+    fn add_positive(&mut self, db: &'db dyn Db, new_positive: Type<'db>) {
         // TODO `Any`/`Unknown`/`Todo` actually should not self-cancel
-        match ty {
-            Type::Intersection(inter) => {
-                let pos = inter.positive(db);
-                let neg = inter.negative(db);
-                self.positive.extend(pos.difference(&self.negative));
-                self.negative.extend(neg.difference(&self.positive));
-                self.positive.retain(|elem| !neg.contains(elem));
-                self.negative.retain(|elem| !pos.contains(elem));
+        if let Type::Intersection(other) = new_positive {
+            for pos in other.positive(db) {
+                self.add_positive(db, *pos);
             }
-            _ => {
-                if !self.negative.remove(&ty) {
-                    self.positive.insert(ty);
-                };
+            for neg in other.negative(db) {
+                self.add_negative(db, *neg);
+            }
+        } else {
+            // ~Literal[True] & bool = Literal[False]
+            if let Type::Instance(class_type) = new_positive {
+                if class_type.is_known(db, KnownClass::Bool) {
+                    let mut found_bool_literal = None;
+                    for neg in &self.negative {
+                        if let Type::BooleanLiteral(bool) = neg {
+                            found_bool_literal = Some(*bool);
+                        }
+                    }
+
+                    if let Some(bool) = found_bool_literal {
+                        *self = Self::new();
```

Comment:
> Since either one is easy to do, which is cheaper, replacing `self` entirely or clearing `self.positive` and `self.negative`?

In an earlier version on this branch, I used `self.positive.clear(); self.negative.clear();` everywhere, but then I changed it to `*self = Self::new()`. Not because it's shorter, but because it seems safer to me. If someone modifies this `struct` to hold additional members, or wants to enforce additional invariants, they don't have to make sure to reset state properly in N different places.

> Does the former have to allocate a new pair of `IndexSet` on the heap, and the latter not? (Maybe creating an empty `IndexSet` doesn't allocate anything on the heap until you add the first element, in which case we'd really only allocate once here.)

For what it's worth, [`OrderMap::new()`](https://docs.rs/ordermap/latest/ordermap/map/struct.OrderMap.html#method.new) does not allocate. It's still true that using `.clear()` twice might be more efficient here because we could re-use the existing allocations when inserting new elements (later).